### PR TITLE
Mentor System Rework (WIP)

### LINKED ISSUE #127: The follow icon on mhelp messages doesn't work
Basically, you know how theres a "F" next to nearly anything that happens that you can click on and your ghost teleports to it, mhelp messages also have that F, but if you click it nothing actually happens.

### PR DESCRIPTION
**Pull request by francinum**
## About The Pull Request
Just because someone else jumps off a cliff doesn't mean you should too. This extends to quarter-assing a port of a system.

## Why It's Good For The Game
Mentor code is awful. Maybe it should be less awful.

closes #127

## Changelog
:cl:
server: Mentor stuff has changed massively. Don't merge this without making me write a full changelog.
/:cl:

### CODE REVIEW

File: code/_globalvars/lists/mobs.dm [hunk 1]
```diff
@@ -1,8 +1,7 @@
 GLOBAL_LIST_EMPTY(clients)							//all clients
-GLOBAL_LIST_EMPTY(admins)							//all clients whom are admins
+GLOBAL_LIST_EMPTY_TYPED(admins, /datum/admins)							//all clients whom are admins
```
Comment (francinum) [inline code/_globalvars/lists/mobs.dm]: There's probably a better solution than making these typed but it gets dreamchecker to shut up.

File: code/modules/admin/holder2.dm [hunk 2]
```diff
@@ -115,7 +115,12 @@ GLOBAL_PROTECT(href_token)
 		owner.holder = src
 		owner.add_admin_verbs()	//TODO <--- todo what? the proc clearly exists and works since its the backbone to our entire admin system
```
Comment (ike709) [inline code/modules/admin/holder2.dm:116]: wew

File: code/modules/admin/verbs/adminhelp.dm [hunk 3]
```diff
@@ -498,24 +521,48 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
...
-	AddInteraction("red", msg, initiator_key_name, claimed_admin_key_name)
-	log_admin_private("Ticket #[id]: [key_name(initiator)]: [msg]")
+	if(!reclassed)
+		AddInteraction("red", msg, initiator_key_name, claimed_admin_key_name)
+		log_admin_private("Ticket #[id]([class]): [key_name(initiator)]: [msg]")
+
 
+	var/notifysound
+
+	if(reclassed)
```
Comment (francinum) [inline code/modules/admin/verbs/adminhelp.dm:576]: This stuff also doesn't seem to be working. Though that might be hubris related to holding sound datums.

File: code/modules/admin/verbs/adminhelp.dm [hunk 4]
```diff
@@ -134,6 +145,7 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 	var/datum/admin_help/ticket = GLOB.ahelp_tickets.TicketByID(ticket_id)
 	//Doing action on a ticket claims it
 	var/claim_ticket = CLAIM_DONTCLAIM
+	to_chat(world, "TBUIACT: [action]")
```
Comment (francinum) [inline code/modules/admin/verbs/adminhelp.dm]: TODO: Clear out debug prints.

File: code/modules/admin/verbs/adminhelp.dm [hunk 5]
```diff
@@ -446,15 +492,20 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
...
 		message_admins("[usr] sent a request to interact with the ticket window without sufficient rights. (Requires: R_ADMIN)")
 		log_admin_private("[usr] sent a request to interact with the ticket window without sufficient rights.")
 		return
+	if(!check_rights(R_MENTOR) && class == "mentor")
+		message_mentors("[usr] sent a request to interact with the ticket window without sufficient rights. (Requires: R_MENTOR)")
+		log_admin_private("[usr] sent a request to interact with the ticket window without sufficient rights.")
+		return
 	//Doing action on a ticket claims it
 	var/claim_ticket = CLAIM_DONTCLAIM
+	to_chat(world, "AHUIACT: [action]")
```
Comment (francinum) [inline code/modules/admin/verbs/adminhelp.dm]: TODO: Clear out debug prints.

File: code/modules/admin/verbs/adminhelp.dm [hunk 6]
```diff
@@ -498,24 +564,51 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
...
+	if(!reclassed)
+		AddInteraction("red", msg, initiator_key_name, claimed_admin_key_name)
+		log_admin_private("Ticket #[id]([class]): [key_name(initiator)]: [msg]")
+
 
+	var/notifysound
+
+	if(reclassed)
+		notifysound = 'sound/effects/yeet.ogg'		//yeeting the ticket around
+		to_chat(world, "RC_YEET USR: [usr.name]")
```
Comment (francinum) [inline code/modules/admin/verbs/adminhelp.dm]: TODO: Clear out debug prints.

File: code/__DEFINES/admin.dm [hunk 7]
```diff
@@ -21,14 +21,18 @@
...
 #define R_DBRANKS		(1<<14)
 
+//Beestation Admin Defines
+#define R_MENTOR		(1<<15)
+#define R_TELEPORT		(1<<16)
+
 #define R_DEFAULT R_AUTOADMIN
 
-#define R_EVERYTHING (1<<15)-1 //the sum of all other rank permissions, used for +EVERYTHING
+#define R_EVERYTHING (1<<17)-1 //the sum of all other rank permissions, used for +EVERYTHING //BEESTATION EDIT - Bumped for bitflags.
```
Comment (Crossedfall) [inline code/__DEFINES/admin.dm]: You don't need to do edit comments like this.

File: code/modules/admin/admin_verbs.dm [hunk 8]
```diff
@@ -183,6 +175,26 @@ GLOBAL_PROTECT(admin_verbs_permissions)
 GLOBAL_LIST_INIT(admin_verbs_poll, list(/client/proc/poll_panel))
 GLOBAL_PROTECT(admin_verbs_poll)
 
+//BEE START - MENTOR2 AND TELEPORT FLAG
```
Comment (Crossedfall) [inline code/modules/admin/admin_verbs.dm]: Same as above

File: code/modules/admin/topic.dm [hunk 9]
```diff
@@ -22,10 +22,7 @@
 	if(!CheckAdminHref(href, href_list))
 		return
 
-	if(href_list["ahelp"])
-		if(!check_rights(R_ADMIN, TRUE))
-			return
-
+	if(href_list["ahelp"]) //BEE CHANGE, Tickets should check their own permissions on ui_state.
```
Comment (Crossedfall) [inline code/modules/admin/topic.dm]: ^

File: code/modules/client/client_procs.dm [hunk 10]
```diff
@@ -225,13 +221,20 @@ GLOBAL_LIST_EMPTY(external_rsc_urls)
 	var/connecting_admin = FALSE //because de-admined admins connecting should be treated like admins.
 	//Admin Authorisation
 	holder = GLOB.admin_datums[ckey]
+	//BEESTATION Only R_ADMIN holders are in GLOB.admins, ditto for R_MENTOR and GLOB.mentors, All staff are in GLOB.staff
```
Comment (Crossedfall) [inline code/modules/client/client_procs.dm]: You get the gist

**Review by Crossedfall — CHANGES_REQUESTED** [review]
We are our own codebase, we don't need "Bee edit" lines scattered around the place.

**Comment by FortSumter** [comment]
I'm a mentor. All I ask is you add aghost to mentors. Other than that as far as usability the current system was fine more or less. I'm sure whatever this does will be good but if it doesn't give mentors the ability to aghost to answer a question it's not going to be any kind of improvement us mentors really care about.

**Comment by francinum** [comment]
@FortSumter Never gonna happen. That's outside your privy access. The follow button WILL work if you're already dead though.

The main thing this fixes was the fact that the mentor system was a technical anomaly, being made primarily of ancient ass code stolen from hippiestation.

You also have access to the ticket menu, roughly parallel to that of the admin team.

**Comment by theboy6545** [comment]
speed merge this shit please i beg

**Comment by francinum** [comment]
I'd like this to get in before anymore changes to TGUI.

**Comment by FortSumter** [comment]
Meeeeh.... I was an trialmin for bee for a while. Having a ticket system is nice, but doesn't change much besides maybe have what we mentors do be acknowledged by Crossed and the headmins. mghost (aghost but for mentors) would of been a real game changer, for it would allow us to get the situational context we need to appropriately answer questions that need it. You can't even argue against doing that due to metagaming because aghosting is logged in admin, and people tend to ask questions about their antag roles anyway. The only trouble I see is that some of the valid hunters on LRP are mentors, and while most all of them are professional enough not to use meta game knowledge to kill someone, it would be difficult to prove their innocence if they were the mentor who answered the antags question.

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Mentor panel should be removed, since all it does now is create SQL errors.

https://github.com/BeeStation/BeeStation-Hornet/blob/master/code/modules/admin/verbs/mentors_edit.dm

File: beestation.dme [hunk 11]
```diff
@@ -2076,16 +2076,7 @@
...
-#include "code\modules\mentor\follow.dm"
-#include "code\modules\mentor\mentor.dm"
-#include "code\modules\mentor\mentor_admintools.dm"
-#include "code\modules\mentor\mentor_memo.dm"
-#include "code\modules\mentor\mentor_verbs.dm"
-#include "code\modules\mentor\mentorhelp.dm"
-#include "code\modules\mentor\mentorpm.dm"
-#include "code\modules\mentor\mentorsay.dm"
-#include "code\modules\mentor\misc_hippiementors.dm"
+#include "code\modules\mentor\mentor2.dm"
```
Comment (Crossedfall) [inline beestation.dme:2079]: Is this listed as `mentor2` for iterative purposes? Or..?

File: code/modules/admin/admin_verbs.dm [hunk 12]
```diff
@@ -183,6 +175,26 @@ GLOBAL_PROTECT(admin_verbs_permissions)
...
+	return list(
+	/client/proc/jumptocoord,			/*we ghost and jump to a coordinate*/
+	/client/proc/jumptoarea,
+	/client/proc/jumptokey,				/*allows us to jump to the location of a mob with a certain ckey*/
+	/client/proc/jumptomob,				/*allows us to jump to a specific mob*/
+	/client/proc/jumptoturf,			/*allows us to jump to a specific turf*/
+	/client/proc/jump_to_ruin
+	)
+
+//BEE END
```
Comment (Crossedfall) [inline code/modules/admin/admin_verbs.dm:196]:
```suggestion
```

**Comment by Crossedfall** [comment]
`Mentor` flag grants access to the Player Panel and subsequently the Banning Panel. While a mentor cannot ban players, they can, however, view their IP & CID using this method. This is a critical blocker.

**Comment by Crossedfall** [comment]
Lots of administrator references still show up for mentor specific things. For example, when a mentor claims a ticket, the end-user sees:
image

If a ticket is rejected they see: 
image

Mentors marking a ticket as IC (Is that a needed aspect of mentor tickets?):
image

Mentors have a Deadmin/Readmin option which incorrectly logs them as dropping/gaining admin perms (minor gripe)

---

Ahelps and Mhelps are linked. Which means if a user has an active mhelp, they wont be able to submit an ahelp and vice versa.

--- 

Clicking the `Transfer` button within the ticket does not send the redirected notice to the player (redirecting via the ticket browser does, however)
image

---

Mentorhelp should arguably be in the Mentor tab. If not, then the mentor tab should just be removed since there is only one option there now, and it's an alias for Staffwho

---

No confirmation is provided to the player when sending an mhelp

**Comment by francinum** [comment]
...I'm not gonna lie I'm *super* tired of touching this branch.

**Comment by TeomanTheGreat** [comment]
When will this be done

**Comment by ike709** [comment]
Fran's busy and doesn't wanna work on this anyways. Someone else can finish it.

**Pull request closed without merging**